crates/ruff_python_parser/src/parser/mod.rs

```diff
@@ -632,6 +632,34 @@ impl<'src> Parser<'src> {
 
         false
     }
+
+    /// Creates a checkpoint to which the parser can later return to using [`Self::rewind`].
+    fn checkpoint(&self) -> ParserCheckpoint<'src> {
+        ParserCheckpoint {
+            tokens: self.tokens.checkpoint(),
+            errors_position: self.errors.len(),
+            current_token_id: self.current_token_id,
+            prev_token_end: self.prev_token_end,
+            recovery_context: self.recovery_context,
+        }
+    }
+
+    /// Restore the parser to the given checkpoint.
+    fn rewind(&mut self, checkpoint: ParserCheckpoint<'src>) {
+        self.tokens.rewind(checkpoint.tokens);
+        self.errors.truncate(checkpoint.errors_position);
+        self.current_token_id = checkpoint.current_token_id;
+        self.prev_token_end = checkpoint.prev_token_end;
+        self.recovery_context = checkpoint.recovery_context;
```

Comment:
Nit: I like to use destructuring here. It adds a compile time guarantee that all fields are used (at the cost of a bit more code)

```rust
let ParserCheckpoint { 
	tokens, 
	errors_position,
	prev_token_end,
	recovery_context
} = checkpoint; 

self.tokens.rewind(tokens);
self.errors.truncate(errors_position);
...
```

You can also do the destructuring right in the method signature, although it then becomes a bit hard to read.

Reply:
Yeah, I recently realized that with the amount of fields increasing in the lexer checkpoint and will be making the changes in a follow-up PR.